xonsh/jobs.py

```diff
@@ -0,0 +1,199 @@
+"""
+A class representing a Python function to be run as a subprocess command.
+"""
+
+try:
+    _shell_tty = sys.stderr.fileno()
+except OSError:
+    _shell_tty = None
+
+_shell_pgrp = os.getpgrp()
+
+
+def _clear_dead_jobs():
+    to_remove = set()
+    for num, job in builtins.__xonsh_all_jobs__.items():
+        obj = job['obj']
+        if isinstance(obj, ProcProxy) or obj.poll() is not None:
+            to_remove.add(num)
+    for i in to_remove:
+        del builtins.__xonsh_all_jobs__[i]
+        if builtins.__xonsh_active_job__ == i:
+            builtins.__xonsh_active_job__ = None
+    if builtins.__xonsh_active_job__ is None:
+        _reactivate_job()
+
+
+def _reactivate_job():
+    if len(builtins.__xonsh_all_jobs__) == 0:
+        return
+    builtins.__xonsh_active_job__ = max(builtins.__xonsh_all_jobs__.items(),
+                                        key=lambda x: x[1]['started'])[0]
+
+
+_block_when_giving = [signal.SIGTTOU, signal.SIGTTIN,
+                      signal.SIGTSTP, signal.SIGCHLD]
+
+
+def _give_terminal_to(pgid):
+    # over-simplified version of:
+    #    give_terminal_to from bash 4.3 source, jobs.c, line 4030
+    # this will give the terminal to the process group pgid
+    if _shell_tty is not None:
+        oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
+        os.tcsetpgrp(_shell_tty, pgid)
+        signal.pthread_sigmask(signal.SIG_SETMASK, oldmask)
+
+
+def print_one_job(num):
+    """
+    Print a line describing job number ``num``.
+    """
+    job = builtins.__xonsh_all_jobs__[num]
+    act = '*' if num == builtins.__xonsh_active_job__ else ' '
+    status = job['status']
+    cmd = [' '.join(i) if isinstance(i, list) else i for i in job['cmds']]
+    cmd = ' '.join(cmd)
+    pid = job['pids'][-1]
+    bg = ' &' if job['bg'] else ''
+    print('{}[{}] {}: {}{} ({})'.format(act, num, status, cmd, bg, pid))
+
+
+def get_next_job_number():
+    """
+    Get the lowest available unique job number (for the next job created)
+    """
+    _clear_dead_jobs()
+    i = 1
+    while i in builtins.__xonsh_all_jobs__:
+        i += 1
+    return i
+
+
+def _default_sigint_handler(num, frame):
+    raise KeyboardInterrupt
+
+
+def wait_for_active_job():
+    """
+    Wait for the active job to finish, to be killed by SIGINT, or to be
+    suspended by ctrl-z.
+    """
+    _clear_dead_jobs()
+    act = builtins.__xonsh_active_job__
+    if act is None:
+        return
+    job = builtins.__xonsh_all_jobs__[act]
+    obj = job['obj']
+    if isinstance(obj, ProcProxy):
+        return
+    if job['bg']:
+        return
+    pgrp = job['pgrp']
+    obj.done = False
+
+    def handle_sigchld(num, frame):
+        obj.done = True
+        job['bg'] = True
+        job['status'] = 'stopped'
+
+    _give_terminal_to(pgrp)  # give the terminal over to the fg process
+    signal.signal(signal.SIGCHLD, handle_sigchld)
+    while obj.poll() is None and not obj.done:
+        time.sleep(0.01)
```

Comment:
Yep.  I think we could probably get away with as much as .2 seconds without it bothering anybody (at least, it didn't bother me, even when I was explicitly looking for it).  I'm not sure how resource-hungry `Popen.poll` is, so I'm not sure how much we need to cut that down.  I'll change to 0.1s for now.